core/vm/eips.go

```diff
@@ -215,8 +215,12 @@ func opTstore(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]b
 
 // opBaseFee implements BASEFEE opcode
 func opBaseFee(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
-	baseFee, _ := uint256.FromBig(interpreter.evm.Context.BaseFee)
-	scope.Stack.push(baseFee)
+	if interpreter.evm.Config.NoBaseFee {
+		scope.Stack.push(new(uint256.Int))
```

Comment:
Instead of checking for `NoBaseFee` in the production opcode, can't we just set the `interpreter.evm.Context.BaseFee` to a `new(uint256.Int)` in the `NoBaseFee`-situation?

Reply:
I did consider that too. The only catch is that some code paths feed us the context directly, and we're not copying currently, so setting the basefee would modify the passed in context. We'd need to add some copying to avoid that.

Reply:
Like this 
```diff
diff --git a/core/vm/eips.go b/core/vm/eips.go
index 44d4ba9cff..35f0a3f7c2 100644
--- a/core/vm/eips.go
+++ b/core/vm/eips.go
@@ -215,12 +215,8 @@ func opTstore(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]b
 
 // opBaseFee implements BASEFEE opcode
 func opBaseFee(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
-	if interpreter.evm.Config.NoBaseFee {
-		scope.Stack.push(new(uint256.Int))
-	} else {
-		baseFee, _ := uint256.FromBig(interpreter.evm.Context.BaseFee)
-		scope.Stack.push(baseFee)
-	}
+	baseFee, _ := uint256.FromBig(interpreter.evm.Context.BaseFee)
+	scope.Stack.push(baseFee)
 	return nil, nil
 }
 
diff --git a/core/vm/evm.go b/core/vm/evm.go
index 2c6cc7d484..1f84ff7660 100644
--- a/core/vm/evm.go
+++ b/core/vm/evm.go
@@ -133,6 +133,9 @@ func NewEVM(blockCtx BlockContext, txCtx TxContext, statedb StateDB, chainConfig
 		chainConfig: chainConfig,
 		chainRules:  chainConfig.Rules(blockCtx.BlockNumber, blockCtx.Random != nil, blockCtx.Time),
 	}
+	if config.NoBaseFee {
+		evm.Context.BaseFee = new(big.Int)
+	}
 	evm.interpreter = NewEVMInterpreter(evm)
 	return evm
 }
```

Reply:
I'll switch to Martin's approach, but we need to be aware that BaseFee form the context is used in many locations, each potentially becoming a footgun with 0.

Reply:
@rjl493456442 Wrt the blob fee, I think we can piggie-back on the same flag. BaseFee, BlobBaseFee, the concept is the same and setting it to zero should also be the same.

Reply:
Ugh, no, Martin's fix isn't good (neither is mine).

Currently the logic is that we only disable base fee checks if 1) NoBaseFee is set and 2) gas prices are 0. Both our suggested fixes unilaterally nuke it out always, which is bad.

Martin's proposal cannot work because you don't know during construction time if  tx will be run with or without gas price. My solution could work if I pulled the gas price up too, but it's a bit ugh. Maybe a cleaner solution is to set/rest the field before actually executing the bytecode for each tx individually. .Will look into that.

Reply:
Maybe Martin's does work. The EVM is reconstructed for each tx.